operator/pkg/nodepools/pools.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package nodepools exists currently because it is required to be a separate due to import conflicts.
+package nodepools
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	vectorizedv1alpha1 "github.com/redpanda-data/redpanda-operator/operator/api/vectorized/v1alpha1"
+	"github.com/redpanda-data/redpanda-operator/operator/pkg/labels"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+func GetNodePools(ctx context.Context, cluster *vectorizedv1alpha1.Cluster, k8sClient client.Reader) ([]*vectorizedv1alpha1.NodePoolSpecWithDeleted, error) {
+	var nodePoolsWithDeleted []*vectorizedv1alpha1.NodePoolSpecWithDeleted
+
+	nps := cluster.GetNodePoolsFromSpec()
+	for i := range nps {
+		np := nps[i]
+
+		nodePoolsWithDeleted = append(nodePoolsWithDeleted, &vectorizedv1alpha1.NodePoolSpecWithDeleted{
+			NodePoolSpec: np,
+		})
+	}
+
+	// Also add "virtual NodePools" based on StatefulSets found.
+	// These represent deleted NodePools - they will not show up in spec.NodePools.
+	var stsList appsv1.StatefulSetList
+	err := k8sClient.List(ctx, &stsList, &client.ListOptions{
+		LabelSelector: labels.ForCluster(cluster).AsClientSelector(),
+	})
+	if err != nil {
+		return nil, err
+	}
+outer:
+	for i := range stsList.Items {
```

Comment:
Looks like this would be much more readable if it was refactored to not be `O(S*N)`. Based on the string checks on line 61, you could index the STSs by their name and then loop over node pools and do a key lookup which should eliminate the check on 71.

Reply:
to be honest, i am not convinced by either the ergonomics of this refactoring, nor the complexity. these are slices with < 10 items, nobody is going to have hundreds or thousands of node pools. a lookup on a map is not much faster - if faster at all, if the slice has only very little items. an array index lookup is always faster than a map lookup (direct memory access vs extra compute by map, with involves hashing, separate chaining or whatever the implementation does), it only wins with a sufficiently large number of N. and N is going to be super small all the time.

Reply:
I think I was misunderstanding how much loop nesting was going on here 😓